agent/src/util/discovery_configuration_controller.rs

```diff
@@ -0,0 +1,638 @@
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+    time::Duration,
+};
+
+use akri_shared::{
+    akri::{
+        configuration::{Configuration, DiscoveryProperty},
+        instance::Instance,
+    },
+    k8s::crud::IntoApi,
+};
+use futures::StreamExt;
+use tokio::sync::mpsc;
+
+use crate::discovery_handler_manager::{
+    discovery_handler_registry::DiscoveryHandlerRegistry, DiscoveryError,
+};
+
+use kube::{Resource, ResourceExt};
+use kube_runtime::{
+    controller::Action,
+    reflector::{ObjectRef, Store},
+    Controller,
+};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error(transparent)]
+    DiscoveryError(#[from] DiscoveryError),
+    #[error(transparent)]
+    Other(#[from] anyhow::Error),
+}
+
+pub trait DiscoveryConfigurationKubeClient: IntoApi<Configuration> + IntoApi<Instance> {}
+
+impl<T: IntoApi<Configuration> + IntoApi<Instance>> DiscoveryConfigurationKubeClient for T {}
+
+pub struct ControllerContext {
+    pub instances_cache: Store<Instance>,
+    pub dh_registry: Arc<dyn DiscoveryHandlerRegistry>,
+    pub client: Arc<dyn DiscoveryConfigurationKubeClient>,
+    pub agent_instance_name: String,
```

Comment:
instance is an overloaded term. Maybe?
```suggestion
    pub agent_identifier: String,
```